source/server/overload_manager_impl.h

```diff
@@ -52,6 +52,53 @@ class OverloadAction {
   Stats::Gauge& scale_percent_gauge_;
 };
 
+// Simple table that converts strings into Symbol instances. Symbols are guaranteed to start at 0
+// and be indexed sequentially.
+class NamedOverloadActionSymbolTable {
```

Comment:
I'm a bit late to this, but did we consider using the stats symbol table code somehow? It's very similar? cc @jmarantz @akonradi

Reply:
I looked briefly at the stats symbol table code but opted not to use it because it has a lot of extra complexity that wasn't needed here. The code here doesn't have to worry about concurrent modifications since the table is built up once at config ingestion time and then treated as read-only for the rest of the life of the process. It also can be append-only since it doesn't need to handle deletions. Maybe a better name would have been a lookup table since that's really all it is.

Reply:
I don't feel super strongly about it but this feels like a bit of a premature optimization and if we feel strongly that it should be optimized we might as well use the "extra complex" thing that is used all over the place.

Reply:
I could see it as a little premature. I've been prototyping how scaled timers fit into the thread-local overload state and I was hoping to head off arguments around the cost of repeated string comparisons [in this code](https://github.com/akonradi/envoy/commit/151f1ca8eb75ad6aa450d62b45d8d12a763659fd#diff-1f8dc4ff3b4aaacce450e5c02a3754ecR136). Representing action names as indices makes things much less expensive.

Reply:
Since I didn't address your second point: my objections to using the stats symbol table were around the performance cost of the extra functionality. Stat names, for instance, are hierarchical, it makes sense to build them up in a tree to save on memory. Overload action names may be notionally hierarchical but they are only ever referenced by their full names, so they can be treated as a single symbol.

Reply:
IMO the question isn't about the complexity of the symbol-table implementation; it's whether the interface is usable. The fact that it has the extra hierarchy doesn't really affect its usage as far as I can tell; you can just put whatever string in and you'll get the same string out.

What you have is probably a little faster though because of no hierarchy or mutex locking. I think the "premature optimization" argument is not whether to do something like a symbol-table, it's whether or not the long-term maintenance of a custom symbol-table is justified by the extra speed.